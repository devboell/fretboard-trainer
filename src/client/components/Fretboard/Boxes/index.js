import React from 'react'


const Boxes = () => (<g />)

Boxes.propTypes = {
  // tuning: pt.arrayOf(pt.string).isRequired,
  // nrOfFrets: pt.number.isRequired,
}

export default Boxes
