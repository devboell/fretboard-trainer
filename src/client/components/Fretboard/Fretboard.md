
default fretboard

```js

<Fretboard
  tuning={['E2', 'A2', 'D3', 'G3', 'B3', 'E4']}
  tuning={12}
/>
```