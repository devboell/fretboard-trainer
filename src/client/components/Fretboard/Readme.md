
### Selections
#### single note selection
```js
<Fretboard
  selectedNotes={['C4']}
/>
```
#### multiple pc selection
```js
<Fretboard
  skinType="strings"
  showSelection
  selectedNotes={['C', 'F']}
/>
```
#### multiple noteObj selection
```js
<Fretboard
  showSelection
  selectedNotes={[{ note: 'C4', label: 'root' }, { note: 'F4', label: '4P' }]}
/>
```
#### single locObj selection, with octaves
```js
<Fretboard
  skinType="strings"
  showOctaves
  showSelection
  selectedLocations={[{ loc: { str: 2, pos: 5 }, label: 'C4' }]}
/>
```

### No Selections
#### default fretboard
```js
<Fretboard
/>
```
```js
<Fretboard
  skinType="strings"
/>
```
#### showNotes
```js
<Fretboard
  showNotes
/>
```
```js
<Fretboard
  skinType="strings"
  showNotes
/>
```
#### showOctaves
```js
<Fretboard
  showNotes
  showOctaves
/>
```
```js
<Fretboard
  skinType="strings"
  showNotes
  showOctaves
/>
```

