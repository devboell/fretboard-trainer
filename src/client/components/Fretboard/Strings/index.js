import React from 'react'
import pt from 'prop-types'
import { times } from 'ramda'
import { stringCenter, fretOffset } from 'lib/fretboard'


const stringLine = nrOfStrings => (str) => {
  const y = stringCenter(nrOfStrings)(str)
  return (
    <line
      key={`str-${str}`}
      x1="0%"
      x2="100%"
      y1={`${y}%`}
      y2={`${y}%`}
    />)
}

const fretLineBounds = nrOfStrings => (
  {
    top: stringCenter(nrOfStrings)(0),
    bottom: stringCenter(nrOfStrings)(nrOfStrings - 1),
  }
)
const fretLine = (nrOfFrets, nrOfStrings) => (frt) => {
  const { top, bottom } = fretLineBounds(nrOfStrings)
  const x = fretOffset(nrOfFrets)(frt)

  return (
    <line
      key={`fret-${frt}`}
      x1={`${x}%`}
      x2={`${x}%`}
      y1={`${top}%`}
      y2={`${bottom}%`}
    />)
}

const Strings = ({ tuning, nrOfFrets }) => {
  const nrOfStrings = tuning.length
  return (
    <g>
      {times(stringLine(nrOfStrings), nrOfStrings)}
      {times(fretLine(nrOfFrets, nrOfStrings), nrOfFrets)}
    </g>
  )
}

Strings.propTypes = {
  tuning: pt.arrayOf(pt.string).isRequired,
  nrOfFrets: pt.number.isRequired,
}

export default Strings
