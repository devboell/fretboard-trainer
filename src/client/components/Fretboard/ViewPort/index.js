import React from 'react'
import { node } from 'prop-types'

const ViewPort = ({
  children,
}) =>
  <svg
    xmlns="http://www.w3.org/2000/svg"
    version="1.1"
    width="100%"
    height="100%"
    stroke="black"
    strokeWidth="1"
    fill="white"
    shapeRendering="geometricPrecision"
    style={{ overflow: 'visible' }}
  >
    {children}
  </svg>

ViewPort.propTypes = {
  children: node.isRequired,
}

export default ViewPort
