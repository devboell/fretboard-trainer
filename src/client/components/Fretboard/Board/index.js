import React from 'react'
import pt from 'prop-types'

import ViewPort from 'components/Fretboard/ViewPort'

import BoxesGraphic from './BoxesGraphic'
import StringsGraphic from './StringsGraphic'
import BoardPositions from './BoardPositions'

const boardGraphicTypes = {
  boxes: BoxesGraphic,
  strings: StringsGraphic,
}

const Board = ({
  tuning,
  nrOfFrets,
  type,
  width,
  offset,
}) => {
  const BoardGraphic = boardGraphicTypes[type]

  return (
    <ViewPort {...{ width, offset }}>
      <BoardGraphic {...{ tuning, nrOfFrets }} />
      <BoardPositions {...{ tuning, nrOfFrets }} />
    </ViewPort>
  )
}

Board.propTypes = {
  tuning: pt.arrayOf(pt.string).isRequired,
  nrOfFrets: pt.number.isRequired,
  type: pt.string.isRequired,
  width: pt.number.isRequired,
  offset: pt.number.isRequired,
}

export default Board
