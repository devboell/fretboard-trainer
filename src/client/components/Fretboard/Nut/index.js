import React from 'react'
import pt from 'prop-types'

import ViewPort from 'components/Fretboard/ViewPort'

const Nut = ({ width, offset }) =>
  <ViewPort {...{ width, offset }} />

Nut.propTypes = {
  width: pt.number.isRequired,
  offset: pt.number.isRequired,
}

export default Nut
