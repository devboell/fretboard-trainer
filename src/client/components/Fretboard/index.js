import React from 'react'
import pt from 'prop-types'
import ViewPort from 'components/ViewPort'

import Strings from './Strings'
import Boxes from './Boxes'

const boardTypes = {
  boxes: Boxes,
  strings: Strings,
}

const board = (tuning, nrOfFrets, type) => {
  const Board = boardTypes[type]
  return <Board {...{ tuning, nrOfFrets }} />
}

const Fretboard = ({ tuning, nrOfFrets, type }) =>
  <ViewPort>
    {board(tuning, nrOfFrets, type)}
  </ViewPort>


Fretboard.propTypes = {
  tuning: pt.arrayOf(pt.string).isRequired,
  nrOfFrets: pt.number.isRequired,
  type: pt.string,
}

Fretboard.defaultProps = {
  type: 'boxes',
}

export default Fretboard
