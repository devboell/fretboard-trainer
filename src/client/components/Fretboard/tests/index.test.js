import React from 'react'

import Fretboard from '../index'

const props = {
  tuning: ['E2', 'A2', 'D3', 'G3', 'B3', 'E4'],
  nrOfFrets: 12,
}

it('Fretboard component, snapshot', () => {
  const wrapper = shallow(<Fretboard {...props} />)
  expect(wrapper).toMatchSnapshot()
})
