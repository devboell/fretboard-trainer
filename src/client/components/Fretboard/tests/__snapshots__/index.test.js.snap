// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fretboard component, snapshot 1`] = `
<ViewPort>
  <Boxes
    nrOfFrets={12}
    tuning={
      Array [
        "E2",
        "A2",
        "D3",
        "G3",
        "B3",
        "E4",
      ]
    }
  />
</ViewPort>
`;
